core/src/main/java/io/questdb/cairo/TableWriter.java

```diff
@@ -874,6 +879,61 @@ public boolean removePartition(long timestamp) {
         }
     }
 
+    public boolean removePartition(Function function) {
+        if (partitionBy == PartitionBy.NONE) {
+            return false;
+        }
+
+        long activePartition = timestampFloorMethod.floor(maxTimestamp);
+        findAllPartitions();
+        int partitionCount = partitionListByTimestamp.size();
+        boolean droppingActivePartition = false;
+        for (int i = 0; i < partitionCount; i++) {
+            long timestampToRemove = partitionListByTimestamp.get(i);
+            dropPartitionFunctionRec.setTimestamp(timestampToRemove);
+            if (function.getBool(dropPartitionFunctionRec)) {
+                partitionsToDrop.add(timestampToRemove);
+                if (activePartition == timestampToRemove) {
+                    droppingActivePartition = true;
+                }
+            }
+        }
+
+        int dropPartitionCount = partitionsToDrop.size();
+        if (dropPartitionCount == 0) {
+            return false;
+        } else if (dropPartitionCount == partitionCount) {
+            truncate();
+            return true;
+        } else if (droppingActivePartition) {
+            LOG.error()
+                    .$("cannot remove active partition [path=").$(path)
+                    .$(", maxTimestamp=").$ts(maxTimestamp)
+                    .$(']').$();
+            return false;
+        } else {
+            for (int i = 0; i < dropPartitionCount; i++) {
+                removePartition(partitionsToDrop.get(i));
+            }
+            return true;
+        }
+    }
+
+    private void findAllPartitions() {
```

Comment:
can we add partitions to "toDrop" list as we find them?

This way we only need one new list.